Adaptive Content Block: Send "Unit viewed" event, selectively display review questions

itsjeyd opened the pull request:
This PR extends the new Adaptive Content Block in two ways:

* It adds logic for notifying Domoscio that the parent unit of an ACB has been accessed by a learner.
* It changes the way child blocks are selected for display: Instead of selecting `max_count` children randomly, ACBs now query Domoscio for information about pending reviews to determine which children to show (if any).

The PR also adds a new course-wide setting called "Adaptive Learning Configuration" which can be used to specify necessary information for communicating with the Domoscio API.

cf. [OC-2150](https://tasks.opencraft.com/browse/OC-2150), [OC-2151](https://tasks.opencraft.com/browse/OC-2151)

**Implementation**

When sending a "Unit viewed" event, ACB assumes that the corresponding unit exists in Domoscio's database (in the form of a "knowledge node" whose `uid` is set to the `block_id` of the unit). This is a safe assumption to make because Domoscio staff will take care of creating entities representing relevant elements of the course structure (including units and review questions) prior to the start date of the course that will be using the new adaptive learning features.

In addition to knowledge nodes representing units, there are other entities that must exist in the Domoscio database to be able to send an event of the type that Domoscio expects: Events target objects that link students to specific knowledge nodes. This means that prior to sending a "Unit viewed" event, the Domoscio database must contain both a "student" entity corresponding to the learner viewing the unit in the LMS, and an object that links the student to the unit (knowledge node). In Domoscio terminology, an object linking a student to a knowledge node is called "knowledge node student". (The plural is "knowledge node students".) Since we can't assume that students and objects that link them to units will be created by Domoscio staff prior to the start date of the course, ACBs will create student and "knowledge node student" objects via the Domoscio API prior to sending "Unit viewed" events (if necessary).

As of this PR, ACBs will also create "knowledge node student" objects that link students to review questions when a learner accesses a unit containing an ACB. This is not a prerequisite for sending "Unit viewed" events (or for obtaining a list of review questions to display from Domoscio). However, when a learner submits an answer to a review question, we will need to notify Domoscio about the result (OC-2152). At that point, links between students and review questions need to exist so we can target them with appropriate `EventResult` events.

**Test instructions**

1. Follow instructions on the [epic](https://tasks.opencraft.com/browse/OC-2127) to set up a FUN box.

2. Switch to this branch (`adaptive-content-block`) and install the new block via

    ```sh
    pip install -r requirements/edx/local.txt
    ```

3. Create a new content library and add a few problems to it.

4. Create a new course and add `"adaptive_library_content"` to advanced modules.

5. Add an "Adaptive Content Block" that references the library created previously to a unit (optionally modifying the number of components to display to each student), and publish the unit. Do **not** access the unit in the LMS yet.

6. To be able to work with the Domoscio API, you'll need to specify a few course-level settings: In Studio, navigate to the Advanced Settings page for the course ("Paramètres" > "Paramètres avancés") and replace the default value for the "Adaptive Learning Configuration" setting with the following:

    ```json
    "url": "<see comment on internal ticket>",
    "instance_id": <see comment on internal ticket>,
    "access_token": "<see comment on internal ticket>",
    "api_version": "v1"
    ```

7. In a venv that has the `requests` library installed, create a knowledge node that corresponds to the unit you created earlier (`block_id` needs to be a 32-character hash), and review the results:

    ```python
    import requests, json
    response = requests.post("<url>/v1/instances/<instance_id>/knowledge_nodes", headers={"Authorization": "Token token=<access_token>"}, data={"knowledge_graph_id": <see comment on internal ticket>, "name": "Unit node", "uid": "<block_id of unit>"})
    json.loads(response.content)
    ```

    Then create a knowledge node for each of the problems that belong to the library you created earlier. Note that problems belonging to a content library are assigned unique block IDs for each Adaptive Content Block that references them; you'll need to use these context-specific IDs when creating corresponding knowledge nodes. Block IDs for problems are 20-character hashes.

8. Review the list of students that exist on the Domoscio test instance:

   ```python
   response = requests.get("<url>/v1/instances/<instance_id>/students", headers={"Authorization": "Token token=<access_token>"})
   json.loads(response.content)
   ```

9. Review the list of links between students and units ("knowledge node student" objects) on the Domoscio test instance:

    ```python
    response = requests.get("<url>/v1/instances/<instance_id>/knowledge_node_students", headers={"Authorization": "Token token=<access_token>"})
    json.loads(response.content)    
    ```

10. Access the unit that contains the ACB in the LMS. Observe that it does not display any child blocks.

11. Review the list of students and the list of "knowledge node student" objects as described above. Observe that the former contains a new student and the latter contains new "knowledge node student" objects, one for each knowledge node you created earlier. The `knowledge_node_id`/`knowledge_node_uid` and `student_id`/`student_uid` fields of the new "knowledge node student" objects should reference the knowledge nodes created earlier, and the student that was created when you accessed the unit for the first time.

12. Review the list of events that have been recorded for the Domoscio test instance, making sure that it contains an entry that references the "knowledge node student" object linking the unit you created earlier to the student that you used to access the unit in the LMS:

    ```python
    response = requests.get("<url>/v1/instances/<instance_id>/events", headers={"Authorization": "Token token=<access_token>"})
    json.loads(response.content)
    ```

**Reviewers**

- [ ] @e-kolpakov / @pomegranited

**Notes**

<!--
- This PR implements one possible approach for fetching pending reviews from the Domoscio API. Depending on how the course structure will be represented in the Domoscio database (details still pending), we might need to adjust the approach in a follow-up PR.
-->
- There is no way to force Domoscio to schedule a review, so it's not possible to manually test the logic for selectively displaying review questions (except the case where no pending reviews are available).

pomegranited commented on common/lib/xmodule/xmodule/library_content_module.py:
```diff
@@ -371,13 +389,159 @@ def get_child_descriptors(self):
+        Check with external service whether any children of this block
+        are scheduled for review by the current user, and return them.
+        """
+        pending_reviews = self.get_pending_reviews(self.anonymous_user_id)
```
Got `ValueError: No JSON object could be decoded` when viewing the ACB unit in the LMS as a student user.  I tried it with 2 different new student accounts on my devstack, and got the same results.

No error is shown for my staff user though.

Full trace below:

```
2016-12-26 07:34:07,620 ERROR 3795 [django.request] base.py:256 - Internal Server Error: /courses/course-v1:OpenCraft+ACB1+OC-2150/courseware/519f9f58a85c4d2aba6182162b1c79e8/e94c284e0a104b77987a3a872367e9d2/
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/auth/decorators.py", line 22, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 110, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/decorators/cache.py", line 43, in _cache_controlled
    response = viewfunc(request, *args, **kw)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/util/views.py", line 41, in inner
    response = view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/views.py", line 352, in index
    return _index_bulk_op(request, course_key, chapter, section, position)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/views.py", line 528, in _index_bulk_op
    context['fragment'] = section_module.render(STUDENT_VIEW, section_render_context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/core.py", line 166, in render
    return self.runtime.render(self, view, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1780, in render
    return self.__getattr__('render')(block, view_name, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1240, in render
    return super(MetricsMixin, self).render(block, view_name, context=context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 806, in render
    frag = view_fn(context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/seq_module.py", line 189, in student_view
    rendered_child = child.render(STUDENT_VIEW, context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/core.py", line 166, in render
    return self.runtime.render(self, view, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1780, in render
    return self.__getattr__('render')(block, view_name, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1240, in render
    return super(MetricsMixin, self).render(block, view_name, context=context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 806, in render
    frag = view_fn(context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/vertical_block.py", line 46, in student_view
    rendered_child = child.render(STUDENT_VIEW, child_context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/core.py", line 166, in render
    return self.runtime.render(self, view, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1780, in render
    return self.__getattr__('render')(block, view_name, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1240, in render
    return super(MetricsMixin, self).render(block, view_name, context=context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 806, in render
    frag = view_fn(context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/library_content_module.py", line 532, in student_view
    return super(AdaptiveLibraryContentModule, self).student_view(context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/library_content_module.py", line 337, in student_view
    for child in self._get_selected_child_blocks():
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/library_content_module.py", line 329, in _get_selected_child_blocks
    for block_type, block_id in self.selected_children():
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/library_content_module.py", line 497, in selected_children
    selected = self.get_selections_current_user(self.children)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/library_content_module.py", line 514, in get_selections_current_user
    pending_reviews = self.get_pending_reviews(self.anonymous_user_id)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/util/adaptive_learning.py", line 242, in get_pending_reviews
    pending_reviews_user = json.loads(response.content)
  File "/usr/lib/python2.7/json/__init__.py", line 326, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 384, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
[26/Dec/2016 07:34:08] "GET /courses/course-v1:OpenCraft+ACB1+OC-2150/courseware/519f9f58a85c4d2aba6182162b1c79e8/e94c284e0a104b77987a3a872367e9d2/ HTTP/1.1" 500 496062
```

itsjeyd commented:
@pomegranited Thanks for testing the changes.

Regarding the question of why no new "knowledge node student" object got created, I think I know what's going on. Inspecting the list of knowledge nodes available on the Domoscio test instance, I noticed that one of them had its `uid` set to a value starting with `block-v1:OpenCraft+ACB1+OC-2150`. That's the one you created manually when following the test instructions, yes? The block ID to use for the `uid` property is just the last part of that value (the 32-character hash). So if the UID of the knowledge node that represents a unit is set to a value that consists of more than the block ID, looking up the knowledge node fails; and that breaks the assumption that a knowledge node representing the unit already exists in the Domoscio database when a learner accesses the unit for the first time. Can you try creating another knowledge node using just the block ID (`6bd172d8109342b98f06bce6dedf197c`) and visit the unit again as `staff`?

Regarding the `ValueError`, you were seeing that because the code was using an API that doesn't return a meaningful anonymous ID for non-staff users. Fixed in  -- we now compute an anonymous ID that's specific to the parent course *and* the Domoscio instance that contains course data.

pomegranited commented on common/lib/xmodule/xmodule/util/adaptive_learning.py:
```diff
@@ -0,0 +1,270 @@
+        for block_id in block_ids:
+            knowledge_node_student = self.get_or_create_knowledge_node_student(block_id, user_id)
+            knowledge_node_students.append(knowledge_node_student)
+        log.error("KNOWLEDGE_NODE_STUDENTS: {}".format(knowledge_node_students))
```
When I try to view the ACB in the LMS, this line is reporting a 500 error.  It tries 9 times to connect to the domoscio URL, and then outputs:
```
2016-12-28 03:54:13,813 ERROR 9304 [xmodule.util.adaptive_learning] adaptive_learning.py:238 - KNOWLEDGE_NODE_STUDENTS: [{u'status': u'500', u'error': u'Internal Server Error'}, {u'status': u'500', u'error': u'Internal Server Error'}]
```

itsjeyd replied:
@pomegranited Nice catch re: the log statement. It got committed accidentally; I'll remove it.

itsjeyd replied:
@pomegranited Regarding the error you were seeing: As of 23902f824c261086ed7c468506de4981762fc988, the code assumes that Domoscio also knows about review questions associated with an ACB (i.e., each review question must exist as a knowledge node in the Domoscio database). This is based on info I got from Domoscio yesterday. The PR description has updated info and test instructions; sorry for not mentioning this.

pomegranited replied:
Ah lovely, that makes sense, and works as you describe.

But how is the client expected to locate the 20-character problem block ID required for adding the problems as knowledge_nodes?  I could only find them in the HTML, and couldn't see them in Studio anywhere.

itsjeyd replied:
@pomegranited The course API gives you a JSON object representing the course structure, and that includes the necessary info. Once the course structure is finalized, Domoscio will use that data to create necessary entities in their database (since they don't provide full support for edX courses at this stage, they'll be doing this manually).

pomegranited replied:
Cool, thanks for the explanation.

itsjeyd commented on common/lib/xmodule/xmodule/library_content_module.py:
```diff
@@ -371,13 +388,186 @@ def get_child_descriptors(self):
+        Return anonymous ID for current user.
+
+        Note that we can't use `user_service.get_anonymous_user_id` here
+        because it only produces meaningful results for staff users.
```
@pomegranited Not sure, no -- I looked into it a little bit but not enough to get a definitive answer.

pomegranited replied:
FYI From the [discussion on TNL-836|https://openedx.atlassian.net/browse/TNL-836?focusedCommentId=89275&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-89275], it sounds like edX are worried about the security implications of allowing arbitrary users to access anyone's anonymous Id.  So they locked it down to only allow logged-in staff users to get a meaningful result from the function.

However, I bet it would be ok to allow arbitrary users to access their _own_ anonymous ID, but not anyone else's.  If you think it's likely that edX would accept this change, then I'd recommend using it into this branch too, to avoid any changes in anonymous IDs stored in Domoscio.

itsjeyd replied:
@pomegranited Just to make sure, are you suggesting to change the original `get_anonymous_user_id` function somehow? If you could give a bit more info on what you had in mind, that would be great :)

pomegranited replied:
Sorry I wasn't clear, @itsjeyd !

If you think upstream would accept it, I'd recommend changing [`xblock_django.user_service.get_anonymous_user_id`](https://github.com/edx/edx-platform/blob/master/common/djangoapps/xblock_django/user_service.py#L34-L55) to something like this:

```python
    def get_anonymous_user_id(self, username, course_id):
        """
        Get the anonymous user id for a user.
        Args:
            username(str): username of a user.
            course_id(str): course id of particular course.
        Returns:
            A unique anonymous_user_id for (user, course) pair.
            None for Non-staff users, unless the current user matches the requested username.
        """
        current_user = self.get_current_user()
        try:
            user = get_user_by_username_or_email(username_or_email=username)
            if not (current_user.opt_attrs.get(ATTR_KEY_USER_ID) == user.id or 
                      current_user.opt_attrs.get(ATTR_KEY_USER_IS_STAFF):
                return None
        except User.DoesNotExist:
            return None

        course_id = CourseKey.from_string(course_id)
        return anonymous_id_for_user(user=user, course_id=course_id, save=False)
```

Then you could use `user_service.get_anonymous_user_id()` directly like you had before in https://github.com/open-craft/edx-platform/pull/49/commits/4904bf47a72900b09159d552cf3c46ebaf0abe39#diff-5b223f1aa12d0cba135cf45dade27de3R500.

But it's difficult to know what the best course of action is.  It would be a major pain if you had to change the anonymous ID algorithm later to get these changes accepted by upstream.

itsjeyd replied:
@pomegranited Thanks for the details! I'll try and see if I can make that change in the context of [OC-2152](https://tasks.opencraft.com/browse/OC-2152). There is a chance that it won't work because in order to complete that story, it will be necessary to obtain an anonymous user ID at the class level. And I'm not sure if it's possible to talk to the user service at that level (only concrete instances of XBlocks seem to have access to that).

itsjeyd commented on common/lib/xmodule/xmodule/library_content_module.py:
```diff
@@ -371,13 +388,186 @@ def get_child_descriptors(self):
+        has been determined based on information from external service providing adaptive learning features,
+        so we only remove blocks that are no longer valid here.
+        """
+        selected = set(tuple(k) for k in selected)  # set of (block_type, block_id) tuples assigned to this student
```
@pomegranited The `make_selection` method of the parent class also computes `selected` this way. I tried changing it the way you suggested and it broke existing tests (for the parent class). I'm not sure why, and I don't have room to look into this in detail. So I'll leave this as is for now to at least be consistent. (I do agree that it looks like it should be possible to change the code in the way you suggested; was thinking the same thing when I first read the parent class' code.)

pomegranited replied:
No worries 😄  Thanks for the explanation.

itsjeyd commented on common/lib/xmodule/xmodule/library_content_module.py:
```diff
@@ -371,13 +388,186 @@ def get_child_descriptors(self):
+        # `added_block_keys`:
+        # External service determines number of children to show,
+        # so no need to *add* any blocks based on value of `max_count` argument.
+
```
@pomegranited Done, see a8e8dbc7a8121bdc0ff181d5671d0d29acf7450a.

itsjeyd commented on common/lib/xmodule/xmodule/util/adaptive_learning.py:
```diff
@@ -0,0 +1,270 @@
+        """
+        url = self.pending_reviews_url
+        payload = {'student_uid': user_id}
+        response = requests.get(url, headers=self.request_headers, data=payload)
```
@pomegranited Thanks for the detailed report :) I ran out of time for fixing these tests yesterday. They should pass as of a8e8dbc7a8121bdc0ff181d5671d0d29acf7450a.

pomegranited replied:
Yep, they are passing now!  Thanks for the fixes.

itsjeyd commented on common/lib/xmodule/xmodule/util/adaptive_learning.py:
```diff
@@ -0,0 +1,270 @@
+        links = json.loads(response.content)
+        return links
+
+    def create_knowledge_node_student(self, block_id, user_id):
```
@pomegranited Yup, that's a great idea. Done in a8ccc570d444a04520e71a8c1bff2959f65846cc.

pomegranited replied:
Cool, thank you!

itsjeyd commented on common/lib/xmodule/xmodule/library_content_module.py:
```diff
@@ -371,13 +388,186 @@ def get_child_descriptors(self):
+        # which is not available at the class level.
+        selected = self.get_selections_current_user(self.children)
+        block_keys = self.make_selection(selected, self.children, self.max_count, "adaptive")  # pylint: disable=no-member
+
```
@pomegranited I don't think there is a nice way to DRY out the `selected_children` methods (at least not without doing a larger refactoring). For example, what you're suggesting won't work because `self.selected` needs to be updated *after* publishing analytics events: One of the methods that handle publishing events uses `self.selected` to compute information about state as it exists prior to updating that field.

It would be possible to introduce some decorators (e.g. for handling the `hasattr` check). But the selection logic is the most complicated part of this code, and I think we'd be trading readability/understandability for DRYness. Since the `selected_children` methods are also pretty short I think it makes sense to leave them as is for now.

pomegranited replied:
Ah gotcha.  Thanks for the explanation!

itsjeyd commented:
@pomegranited Thanks for the review! I addressed all of your comments, so this is ready for you to have another look. As mentioned in https://github.com/open-craft/edx-platform/pull/49#discussion_r94029148, please see the PR description for up-to-date info and test instructions.

> Would like the functionality for OC-2151 to be finished before I approve the whole PR though, so I can see the questions in the LMS.

Domoscio is in the process of making changes on their end as well, so it might take some time until we start getting pending reviews from the API. To avoid overhead on our end, it would be good to merge this PR once the internal review is done (OC-2152 will need to build on the changes from this PR). We'll have another chance to verify the pending reviews logic in the context of OC-2154.

> I checked for accessibility issues - TODO once questions are shown in LMS.

We won't be making any changes to user interactions involving ACBs (or to the way that children of ACBs are displayed in the LMS) so we should be good here :)

pomegranited commented:
@itsjeyd Your changes look good, and work as described.  I've left a couple of questions, but nothing that requires changes before merging.

* https://github.com/open-craft/edx-platform/pull/49#discussion_r94093392
* https://github.com/open-craft/edx-platform/pull/49#discussion_r94093681

itsjeyd commented:
Thanks @pomegranited, I replied to your questions.

@e-kolpakov Could you approve this PR as well please? Your comments have long been addressed :)

itsjeyd commented:
Thanks @e-kolpakov! Merging now.
